pkg/broker/ingress/multi_topic_decouple_sink.go

```diff
@@ -53,31 +64,84 @@ type multiTopicDecoupleSink struct {
 
 // Send sends incoming event to its corresponding pubsub topic based on which broker it belongs to.
 func (m *multiTopicDecoupleSink) Send(ctx context.Context, ns, broker string, event cev2.Event) protocol.Result {
-	topic, err := m.getTopicForBroker(ns, broker)
+	topic, err := m.getTopicForBroker(types.NamespacedName{Namespace: ns, Name: broker})
 	if err != nil {
 		return err
 	}
-	ctx = cecontext.WithTopic(ctx, topic)
-	return m.client.Send(ctx, event)
+
+	dt := extensions.FromSpanContext(trace.FromContext(ctx).SpanContext())
+	msg := new(pubsub.Message)
+	if err := cepubsub.WritePubSubMessage(ctx, binding.ToMessage(&event), msg, dt.WriteTransformer()); err != nil {
+		return err
+	}
+
+	_, err = topic.Publish(ctx, msg).Get(ctx)
+	return err
 }
 
 // getTopicForBroker finds the corresponding decouple topic for the broker from the mounted broker configmap volume.
-func (m *multiTopicDecoupleSink) getTopicForBroker(ns, broker string) (string, error) {
-	brokerConfig, ok := m.brokerConfig.GetBroker(ns, broker)
+func (m *multiTopicDecoupleSink) getTopicForBroker(broker types.NamespacedName) (*pubsub.Topic, error) {
+	topicID, err := m.getTopicIDForBroker(broker)
+	if err != nil {
+		return nil, err
+	}
+
+	if topic, ok := m.getExistingTopic(broker); ok {
+		// Check that the broker's topic ID hasn't changed.
+		if topic.ID() == topicID {
+			return topic, nil
+		}
+	}
+
```

Comment:
`updateTopicForBroker` does these checks too. It seems to me you can just directly call `updateTopicForBroker`. (maybe rename it to `getOrCreateTopicForBroker`

Reply:
will lgtm once this is addressed.

Reply:
This would require obtaining the mutex on every request. Most requests should only require an RLock.

Reply:
right. In that case is it necessary for `updateTopicForBroker` to do the same checks? How about 
```
func (m *multiTopicDecoupleSink) updateTopicForBroker(broker types.NamespacedName, topicID string) (*pubsub.Topic, error) {
	m.topicsMut.Lock()
	defer m.topicsMut.Unlock()

	if topic, ok := m.topics[broker]; ok {
		// Stop old topic.
		m.topics[broker].Stop()
	}
	topic := m.pubsub.Topic(topicID)
	m.topics[broker] = topic
	return topic, nil
}
```

Reply:
I guess there could be a race condition where the given `topicID` is not up to date.

Reply:
Yes, to avoid potentially creating unnecessary topic connections it's necessary to recheck state after obtaining the mutex. This path will only be hit once per topic so the overall impact should be negligible.